network/p2p/tracer/internal/rpc_sent_tracker.go

```diff
@@ -1,32 +1,130 @@
 type RPCSentTracker struct {
-	cache *rpcSentCache
+	component.Component
+	*lastHighestIHaveRPCSize
+	cache                                *rpcSentCache
+	workerPool                           *worker.Pool[trackableRPC]
+	lastHighestIHaveRPCSizeResetInterval time.Duration
+}
+
+// RPCSentTrackerConfig configuration for the RPCSentTracker.
+type RPCSentTrackerConfig struct {
+	Logger zerolog.Logger
+	//RPCSentCacheSize size of the *rpcSentCache cache.
+	RPCSentCacheSize uint32
+	// RPCSentCacheCollector metrics collector for the *rpcSentCache cache.
+	RPCSentCacheCollector module.HeroCacheMetrics
+	// WorkerQueueCacheCollector metrics factory for the worker pool.
+	WorkerQueueCacheCollector module.HeroCacheMetrics
+	// WorkerQueueCacheSize the worker pool herostore cache size.
+	WorkerQueueCacheSize uint32
+	// NumOfWorkers number of workers in the worker pool.
+	NumOfWorkers int
+	// LastHighestIhavesSentResetInterval the refresh interval to reset the lastHighestIHaveRPCSize.
+	LastHighestIhavesSentResetInterval time.Duration
 }
 
 // NewRPCSentTracker returns a new *NewRPCSentTracker.
-func NewRPCSentTracker(logger zerolog.Logger, sizeLimit uint32, collector module.HeroCacheMetrics) *RPCSentTracker {
-	config := &rpcCtrlMsgSentCacheConfig{
-		sizeLimit: sizeLimit,
-		logger:    logger,
-		collector: collector,
+func NewRPCSentTracker(config *RPCSentTrackerConfig) *RPCSentTracker {
+	cacheConfig := &rpcCtrlMsgSentCacheConfig{
+		sizeLimit: config.RPCSentCacheSize,
+		logger:    config.Logger,
+		collector: config.RPCSentCacheCollector,
 	}
-	return &RPCSentTracker{cache: newRPCSentCache(config)}
+
+	store := queue.NewHeroStore(
+		config.WorkerQueueCacheSize,
+		config.Logger,
+		config.WorkerQueueCacheCollector)
+
+	tracker := &RPCSentTracker{
+		lastHighestIHaveRPCSize:              &lastHighestIHaveRPCSize{atomic.NewInt64(0), time.Now()},
+		cache:                                newRPCSentCache(cacheConfig),
+		lastHighestIHaveRPCSizeResetInterval: config.LastHighestIhavesSentResetInterval,
+	}
+	tracker.workerPool = worker.NewWorkerPoolBuilder[trackableRPC](
+		config.Logger,
+		store,
+		tracker.rpcSentWorkerLogic).Build()
+
+	builder := component.NewComponentManagerBuilder()
+	for i := 0; i < config.NumOfWorkers; i++ {
+		builder.AddWorker(tracker.workerPool.WorkerLogic())
+	}
+	tracker.Component = builder.Build()
+	return tracker
 }
 
-// OnIHaveRPCSent caches a unique entity message ID for each message ID included in each rpc iHave control message.
+// Track submits the control message to the worker queue for async tracking.
 // Args:
 // - *pubsub.RPC: the rpc sent.
-func (t *RPCSentTracker) OnIHaveRPCSent(iHaves []*pb.ControlIHave) {
+// All errors returned from this function can be considered benign.
+func (t *RPCSentTracker) Track(rpc *pubsub.RPC) error {
+	n, err := nonce()
+	if err != nil {
+		return fmt.Errorf("failed to get track rpc work nonce: %w", err)
+	}
+	if ok := t.workerPool.Submit(trackableRPC{Nonce: n, rpc: rpc}); !ok {
```

Comment:
Please add tests for the following scenarios: 
1. Duplicate RPCs with different nonce values; both should be processed. 
2. Distinct RPCs tracked concurrently; a worker should eventually pick each.

Reply:
https://github.com/onflow/flow-go/pull/4553/commits/7c5daef4ab8ea91ab7cb03e51a1820198a9583c5